google/cloud/bigquery_storage_v1/writer.py

```diff
@@ -316,64 +177,51 @@ def close(self, reason: Optional[Exception] = None):
                 an "intentional" shutdown. This is passed to the callbacks
                 specified via :meth:`add_close_callback`.
         """
-        self._shutdown(reason=reason)
-
-    def _shutdown(self, reason: Optional[Exception] = None):
-        """Run the actual shutdown sequence (stop the stream and all helper threads).
-
-        Args:
-            reason:
-                The reason to close the stream. If ``None``, this is
-                considered an "intentional" shutdown.
-        """
-        with self._closing:
-            if self._closed:
-                return
-
-            # Stop consuming messages.
+        with self._thread_lock:
             if self.is_active:
```

Comment:
This is probably just stylistic, but do you want to unconditionally call the Connection's close and have it handle all the internal state checks?  is_active is more of less just peeking into the Connection state indirectly.

You could also consider not raising the duplicate close error if the method is really idempotent, but it might require refinements to the Connection implementations you tackle in a followup PR.

Reply:
> This is probably just stylistic, but do you want to unconditionally call the Connection's close and have it handle all the internal state checks? is_active is more of less just peeking into the Connection state indirectly.

Yeah, I think having AppendRowsStream more decoupled from _Connection will make the code easier to maintain or update in the future.

> You could also consider not raising the duplicate close error if the method is really idempotent, but it might require refinements to the Connection implementations you tackle in a followup PR.

It would be a great idea to ensure Connection's close method is idempotent. I can remove the check once this step is done.